### PR TITLE
Basic Testing

### PR DESCRIPTION
**Pull request by michaelcolomb**
This pull request adds basic unit testing, independent of the server running. The tests are meant to run with a continuous integration, being run with each push. This is a follow-up to #26.

Summary of changes:
- [x] UserService Tests
- [x] TokenService Tests
- [x] StorageService Tests
- [x] PermissionService Tests
- [x] AuthService Tests

Remaining work:

Unit tests for MailService have not been added, because it is currently unconducive to unit testing; it will be refactored in the future so that it is not reliant on development / production status.

### CODE REVIEW

**Comment by nmagerko** [comment]
That was a lot of work -- great job. I'll review this an open an issue for the `MailService` soon.

**Comment by shotaroikeda** [comment]
Note that we initial (and mistakenly) created an end to end tests instead. We left them under `test/_end_to_end` for the future.

**Comment by nmagerko** [comment]
@shotaroikeda Yeah that's fine. @michaelcolomb sorry I haven't gotten to this yet; busy week and this is a big PR. Will get on it soon.

File: README.md [hunk 1]
```diff
@@ -132,6 +132,20 @@ npm run prod
 Use `Control-C` to kill the server. Note that `node` must be on your path and that
 the configuration for the target environment must be present in the `config` directory.
 
+
+If you would like to see the available commands, run
+``` shell
+node api.js -h
```
Comment (nmagerko) [inline README.md]: Why is this necessary?

File: README.md [hunk 2]
```diff
@@ -145,3 +159,12 @@ Any issues found in a feature branch (i.e. not in `master` or `staging`)
 should be communicated to the active developer(s) directly, unless there is an open
 pull-request for that feature branch. In the latter case, just leave a comment on
 the pull-request detailing the issue preventing a merge.
+
+## Testing
+
+Please set the proper enviornment variables before testing. You can run the tests locally with
+
+``` shell
+node api.js -t
```
Comment (nmagerko) [inline README.md]: This should be an `npm` script that can be run via `npm test` or `npm run test`. It should set up the environment variables from a config file and then run the tests

File: api/v1/controllers/AuthController.js [hunk 3]
```diff
@@ -99,12 +99,11 @@ function passwordReset(req, res, next) {
 			return null;
 		});
 }
-
 router.use(bodyParser.json());
 router.use(middleware.auth);
 router.use(middleware.request);
 
-router.post('', createToken);
+router.post('/', createToken);
```
Comment (nmagerko) [inline api/v1/controllers/AuthController.js]: Why was this changed? This looked correct as it was

File: api/v1/services/TokenService.js [hunk 4]
```diff
@@ -55,6 +54,8 @@ module.exports.generateToken = function(user, scope){
 	var tokenVal = utils.crypto.generateResetToken();
 	var userId = user.get('id');
 
+	//Token.where({user_id: userId, type: scope}).fetchAll().then(function(data){console.log(data)});
```
Comment (nmagerko) [inline api/v1/services/TokenService.js]: What is this doing here?

File: test/auth.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+var AuthService = require('../api/v1/services/AuthService.js');
+
+var config = require('../api/config');
+var JWT_SECRET = config.auth.secret;
+var jwt = require('jsonwebtoken');
+
+var assert = chai.assert;
+var expect = chai.expect;
+
+describe('UserService',function(){
```
Comment (nmagerko) [inline test/auth.js]: Wrong service

File: test/auth.js [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+            parameters = {
+                expiresIn: config.auth.expiration,
+                subject: '1'
+            };
+
+            done();
+        });
+        it('issues a token for a valid user',function(done){
+            var token = AuthService.issueForUser(testUser);
+            expect(token).to.eventually.equal(jwt.sign(subject, JWT_SECRET,parameters)).and.notify(done);
```
Comment (nmagerko) [inline test/auth.js]: Let's not make a new token to check the JWT. You should be able to decode the JWT with the library you imported and just check the values of the token directly.

File: test/mocked/User.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
+var mockery = require('mockery');
```
Comment (nmagerko) [inline test/mocked/User.js]: Do we need this file (`test/mocked/User.js`) still? It seems like this was part of the old tests.

File: test/mocked/index.js [hunk 8]
```diff
@@ -0,0 +1 @@
+require('./User');
```
Comment (nmagerko) [inline test/mocked/index.js]: If the `User` mocked object is not needed, get rid of this.

File: test/permission.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+        it('allows creation by SUPERUSER', function(done){
+            var testUser = User.forge({ id: 1, email: 'new@example.com' });
+            testUser.setPassword('password123').then(function () {
+                testUser.related('roles').add({ role: utils.roles.SUPERUSER});
+                var allow = PermissionService.canCreateUser(testUser);
+                expect(allow).to.eventually.equal(true).and.notify(done);
+           });
+        });
+        it('allows creation of COMMON by ORGANIZER', function(done){
+            var testUser = User.forge({ id: 1, email: 'new@example.com' });
```
Comment (nmagerko) [inline test/permission.js]: Try to refactor this test. It looks like it was almost copied and pasted from the test above. It'd be better to make a helper function that returns the texting context `allow` for you given some input parameters.

File: test/permission.js [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+            testUser.setPassword('password123').then(function () {
+                testUser.related('roles').add({ role: "ADMIN"});
+                var allow = PermissionService.canCreateUser(testUser,'MENTOR');
+                expect(allow).to.eventually.equal(true).and.notify(done);
+            });
+        });
+        it('denies creation of COMMON by non-ORGANIZER', function(done){
+            var testUser = User.forge({ id: 1, email: 'new@example.com' });
```
Comment (nmagerko) [inline test/permission.js]: Same comment here. Just call a helper function

File: test/test.js [hunk 11]
```diff
@@ -0,0 +1,9 @@
+var chai = require('chai');
+chai.use(require('chai-as-promised'));
+
+// TODO unit tests go here
```
Comment (nmagerko) [inline test/test.js]: Remove this todo

File: test/token.js [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+        it('finds valid token',function(done){
+            var found = TokenService.findTokenByValue(tokenVal,'DEFAULT');
+            expect(found).to.eventually.have.deep.property("attributes.user_id", 1,"user ID should be 1")
+                .then(function(){
+                    expect(found).to.eventually.have.deep.property("attributes.value", tokenVal,"token value sould be "+tokenVal)
+                        .and.notify(done);
+                });
+        });
+        it('throws error for invalid scope',function(done){
+            // TODO structure from 'user.js' was not working for catching this error; refactor
```
Comment (nmagerko) [inline test/token.js]: Does this still need to be refactored?

File: test/token.js [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+        });
+        it('throws error with expired token and calls delete on token',function(done){
+            var _get = sinon.stub(Token.prototype,'get');
+            _get.withArgs('created').returns(0);
+            var _destroy = sinon.stub(Token.prototype,'destroy');
+            var found = TokenService.findTokenByValue(tokenVal,'DEFAULT');
+            expect(found).to.eventually.be.rejectedWith(errors.TokenExpirationError)
+                .then(function(){
+                    expect(_destroy.neverCalledWith()).to.equal(false);
+                    _get.restore;
```
Comment (nmagerko) [inline test/token.js]: What does `.restore` do? Is it a function?

File: test/user.js [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+
+				done();
+			});
+		});
+		it('creates a new user', function (done) {
+			var user = UserService.createUser('new@example.com', 'password123', utils.roles.ATTENDEE);
+			expect(user).to.eventually.have.deep.property("attributes.id", 1);
+			expect(user).to.eventually.have.deep.property("attributes.email", 'new@example.com');
+
+			// TODO restructure this
```
Comment (nmagerko) [inline test/user.js]: Do we still need to do this? I remember we talked about a better way of making it work.

File: test/user.js [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+				_findById = sinon.stub(User, 'findById');
+
+				_findById.withArgs(1).returns(_Promise.resolve(testUser));
+				_findById.withArgs(sinon.match.number).returns(_Promise.resolve(null));
+
+				done();
+		});
+		it('finds existing user',function(done){
+			var user = UserService.findUserById(1);
+			expect(user).to.eventually.have.deep.property("attributes.id", 1,"ID should 1, the searched for ID")
```
Comment (nmagerko) [inline test/user.js]: Please fix the messages of this `expect` and the one below (add the word 'be').

File: test/token.js (hunk 13 above)
Comment (michaelcolomb) [inline test/token.js]: `.restore` removes the stub from the function. Remember you actually used it in `test/user.js`.

File: api/v1/controllers/AuthController.js (hunk 3 above)
Comment (nmagerko) [inline api/v1/controllers/AuthController.js]: Please change it back.

File: test/token.js (hunk 13 above)
Comment (nmagerko) [inline test/token.js]: Right, so you need to call `.restore()` not `.restore`, correct?

File: config/test.config.template [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+#                               TEST_ENV_CONFIG
+###############################################################################
+#ENV AND PORTS
+NODE_ENV=test
+PROFILE=hackillinois-api
+HACKILLINOIS_PORT=8080
+
+#API KEYS AND PASSWORDS
+HACKILLINOIS_SECRET=
+HACKILLINOIS_SUPERUSER_EMAIL=
```
Comment (shotaroikeda) [inline config/test.config.template]: Do the tests run w/o setting these environments?

File: test/mocked/User.js (hunk 7 above)
Comment (shotaroikeda) [inline test/mocked/User.js]: Yeah I think this belongs in the `_end_to_end/` folder

File: test/storage.js [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+var User = require('../api/v1/models/User.js');
+var client = require('aws-sdk');
+
+var assert = chai.assert;
+var expect = chai.expect;
+
+describe('StorageService',function(){
+
+    describe('createUpload', function () {
+        var _forge;
```
Comment (shotaroikeda) [inline test/storage.js]: Try to avoid having underscores as variable names if possible. They're usually reserved for system-dependent variables within libraries.

Comment (nmagerko) [inline test/storage.js]: Let's allow this for now. This is a sort of system-dependent variable -- if it's modified directly, the tests will fail (until restore is called).
